taichi/ir/frontend_ir.h

```diff
@@ -447,9 +447,16 @@ class IdExpression : public Expression {
   }
 
   void flatten(FlattenContext *ctx) override {
-    ctx->push_back(std::make_unique<LocalLoadStmt>(
-        LocalAddress(ctx->current_block->lookup_var(id), 0)));
-    stmt = ctx->back_stmt();
+    auto var_stmt = ctx->current_block->lookup_var(id);
+    if (var_stmt->is<AllocaStmt>()) {
+      ctx->push_back(
+          std::make_unique<LocalLoadStmt>(LocalAddress(var_stmt, 0)));
+      stmt = ctx->back_stmt();
+    } else {
+      // The loop index may have a coordinate offset.
+      TI_ASSERT(var_stmt->is<LoopIndexStmt>() || var_stmt->is<BinaryOpStmt>());
```

Comment:
IIUC, `BinaryOpStmt` is to support the offset case, right? Personally I feel this is a bit ad-hoc. WDYT if we make `LoopIndexStmt` contain an optional `offset` (maybe in another PR) ? Will that cause lots of works?

Reply:
I think adding an optional `offset` to `LoopIndexStmt` would cause further analysis and optimizations much harder.

Actually it looks good to me if `IdExpression::flatten` results in *any* `Stmt *` in the future. `AllocaStmt`, however, should be the "special case" that we need to load it.

Reply:
Ack, thx!